Obsidian.IO/ProtocolReader.cs

```diff
@@ -0,0 +1,300 @@
+        /// </summary>
+        public ReadOnlySpan<byte> CurrentSpan => new(buffer, index, length - index);
+        
+        
+        /// <summary>
+        /// The reader's current position
+        /// </summary>
+        public int Position
+        {
+            get => index;
+            set => index = value;
+        }
+
+        private readonly byte[] buffer;
+        private readonly int length;
+        private int index;
+        
+
+        /// <summary>
+        /// Creates a <see cref="ProtocolReader"/> using the underlying array of the <see cref="ReadOnlyMemory{T}"/>
+        /// </summary>
+        /// <param name="memory"><see cref="ReadOnlyMemory{T}"/> created using an array</param>
+        /// <exception cref="InvalidOperationException">The provided <see cref="ReadOnlyMemory{T}"/> was not created from an array</exception>
+        public ProtocolReader(ReadOnlyMemory<byte> memory)
+        {
+            if (!MemoryMarshal.TryGetArray(memory, out var segment))
+                throw new InvalidOperationException("Cannot get an array from the ReadOnlyMemory");
+
+            
+            buffer = segment.Array!;
+            length = memory.Length;
+            index = 0;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="ProtocolReader"/> using the provided <see cref="buffer"/> and <see cref="length"/>
+        /// </summary>
+        /// <param name="buffer">Buffer to read from</param>
+        /// <param name="length"><see cref="buffer"/>'s length</param>
+        public ProtocolReader(byte[] buffer, int? length = null)
+        {
+            this.buffer = buffer;
+            this.length = length ?? buffer.Length;
+            index = 0;
+        }
+
+        /// <summary>
+        /// Reads a <see cref="Span{T}"/> to the target <see cref="Span{T}"/>
+        /// </summary>
+        /// <param name="span">The <see cref="Span{T}"/> to write to</param>
+        /// <exception cref="IndexOutOfRangeException">There is not enough data in the reader's <see cref="Buffer"/></exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe void ReadSpan(Span<byte> span)
+        {
+            EnsureSize(span.Length);
+
+            fixed (byte* sptr = span)
+            {
+                fixed (byte* bptr = buffer)
+                {
+                    System.Buffer.MemoryCopy(bptr + index, sptr, span.Length, span.Length);
+                }
+            }
+            index += span.Length;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnsureSize(int required)
+        {
+            if (index + required > length)
+                throw new IndexOutOfRangeException("Cannot read past buffer's end");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public byte ReadByte()
+        {
+            EnsureSize(1);
```

Comment:
Me: Hey mom, can we have `sizeof(byte)`?
Mom: No, we have `sizeof(byte)` at home
`sizeof(byte)` at home: